src/v/cluster/scheduling/partition_allocator.cc

```diff
@@ -108,11 +108,29 @@ allocation_constraints partition_allocator::default_constraints() {
  * with partitions that cannot be re-accommodated on smaller peers).
  */
 std::error_code partition_allocator::check_cluster_limits(
-  const uint64_t new_partitions_replicas_requested) const {
+  const uint64_t new_partitions_replicas_requested,
+  const model::topic_namespace& topic) const {
     if (_members.local().nodes().empty()) {
         // Empty members table, we're probably running in a unit test
         return errc::success;
     }
+    // Internal topics are excluded from checks to prevent allocation failures
+    // when creating them. This is okay because they are fairly small in number
+    // compared to kafka user topic partitions.
+    auto is_internal_ns = topic.ns == model::redpanda_ns
+                          || topic.ns == model::kafka_internal_namespace;
+    const auto& internal_topics = _internal_kafka_topics();
+    auto is_internal_topic = topic.ns == model::kafka_namespace
+                             && std::any_of(
+                               internal_topics.cbegin(),
+                               internal_topics.cend(),
+                               [&topic](const ss::sstring& internal_topic) {
+                                   return internal_topic == topic.tp;
+                               });
+    if (is_internal_ns || is_internal_topic) {
+        return errc::success;
+    }
```

Comment:
https://github.com/redpanda-data/redpanda/blob/60ad2f4305f98e03ab10671f16518bc27ddac978/src/v/cluster/scheduling/allocation_node.cc#L58-L70

would it be possible to abstract these into a function, instead of duplicating the logic?

The first part
```
    auto is_internal_ns = topic.ns == model::redpanda_ns
                          || topic.ns == model::kafka_internal_namespace;
```
seems to be used in a couple more places, on its own. Not sure what the differences are ...

Reply:
> would it be possible to abstract these into a function, instead of duplicating the logic?

I wanted to but eventually gave up as "not worth the effort" because there wasn't an obvious place to put this code to and because the input type in the two places is slightly different `model::ntp` vs `model::topic_namespace`. But I can be convinced to take another stab at this.

> seems to be used in a couple more places, on its own. Not sure what the differences are ...

Yes, the definition of an internal topic is a bit fuzzy across the codebase. Some only consider internal namespaces (`internal` and `kafka_internal`) as internal topics, and some also include the topics in `kafka_nodelete_topics` (which tend to contain the internally created topics, like the audit log topics, or the consumer offsets topic, transactions topic). In the case of partition allocations, what we really want to exclude are topics that are internally created, so combining these two is a fair proxy. This is what is done in `allocation_node::is_full`, so the concept of what is an internal topic inside `src/v/cluster/scheduling` is consistent.

Reply:
+1 to extracting this to a static helper function in `allocation_node` - seems important that we have the same logic in these two places. We can have this function take `topic_namespace_view` as a parameter, then it is usable with an ntp as well.

Reply:
Fixed: https://github.com/redpanda-data/redpanda/pull/24409#issuecomment-2518060547